aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -756,9 +785,101 @@ Tensor dense_to_sparse_bsr(const Tensor& self, IntArrayRef blocksize) {
 }
 
 Tensor dense_to_sparse_bsc(const Tensor& self, IntArrayRef blocksize) {
-  AT_ERROR(
-      "Conversion from ", self.layout(), " to SparseBsc is currently not supported.");
-  return self;
+  TORCH_CHECK(
+      blocksize[0] > 0 && blocksize[1] > 0,
+      "blocksize needs to be non zero, but got ",
+      blocksize);
+  TORCH_CHECK(
+      self.size(-2) % blocksize[0] == 0,
+      "Tensor size(-2) ",
+      self.size(-2),
+      " needs to be divisible by blocksize[0] ",
+      blocksize[0]);
+  TORCH_CHECK(
+      self.size(-1) % blocksize[1] == 0,
+      "Tensor size(-1) ",
+      self.size(-1),
+      " needs to be divisible by blocksize[1] ",
+      blocksize[1]);
+  auto n_batch_dim = self.dim() - 2;
+  auto is_batched = n_batch_dim > 0;
+  auto values = _batch_tile_tensor(self, blocksize);
+  auto not_zero_mask = _batch_tile_tensor((self != 0), blocksize);
+  not_zero_mask = not_zero_mask.any(-1).any(-1);
+
+  if (is_batched) {
+    // Some extra setup and checks for batched
+    if (n_batch_dim > 1) {
+      // Flatten to single batch dim
+      values = values.flatten(0, n_batch_dim - 1);
+      not_zero_mask = not_zero_mask.flatten(0, n_batch_dim - 1);
+    }
+    TORCH_CHECK(
+        not_zero_mask.size(0) > 0,
+        "to_sparse_bsc: Expected product of batch dimensions to be non-zero.");
+
+    // Batched inputs must have the  same nnz per batch.
+    auto not_zero_mask_0 = not_zero_mask.select(0, 0);
+    auto nse_per_batch = not_zero_mask_0.sum().repeat(not_zero_mask.size(0));
+    TORCH_CHECK(
+        not_zero_mask.sum({-2, -1}).equal(nse_per_batch),
+        "Expect the same number of specified elements per batch.");
+
+    // We need to join batches into a matrix increasing the number of cols.
+    // This allows us to create indices for col compressed matrix and de-batch
+    // them later. Avoiding a kernel start for each batch.
+    values = values.transpose(0, 1).flatten(1, 2);
+    not_zero_mask = not_zero_mask.transpose(0, 1).flatten(1, 2);
+  }
```

Comment:
I am not sure I understand what is happening here. Could you please clarify?

Reply:
We have a kernel for generating `compressed_indices` from coo indices, but it only works for matrix inputs. Rather than doing that conversion for each batch I worked out a way to go from 1-d compressed indices (the compressed indices vector for a sparse matrix) to 2-d compressed (the compressed indices for a 3-d/1 batch-dim sparse tensor). 

It works only if the compressed (matrix) indices it is given are for a matrix where the batches are joined such that the compressed axis has the batch dim folded into it. So here we need to get to `(r, b*c)` from `(b, r, c)`, and also have the underlying data in that order.  The first rows of the matrix looks like:

```
[batch_0_row_0, batch_1_row_0,... batch_b-1_row_0],
[batch_0_row_1, batch_1_row_1,... batch_b-1_row_1],
```

If you look at the follow on PRs where we use this for dense->BSR/CSR. There is no transpose because the batch and row dims are already adjacent. So `flatten(0, 1)` does the trick.